web/testing/regression-tests/src/test-runner.js

```diff
@@ -0,0 +1,289 @@
+   * @param {string} locator   "shortname/id", e.g. "k/kayan"
+   * @returns {object}  object with `shortname` and `id` properties
+   */
+  parseLocator: function(locator) {
+    let m = locator.match(/^([a-z0-9_]+)\/([a-z0-9_]+)$/);
+    return m ?
+      {shortname: m[1], id: m[2]} :
+      {shortname: null, id: null};
+  },
+
+  /**
+   * Called by a .tests file to register a set of input tests
+   * @param {object} data  
+   */
+  register: function(data) {
+    let keys = Object.keys(data.inputTests);
+    // We are going to refactor .inputTests from an object into an array with "id" parameters.
+    data.inputTests = keys.map((key) => Object.assign({id: key}, data.inputTests[key]));
+    this.keyboards[data.keyboard.id] = data;
+  },
+
+  /**
+   * Removes a keyboard and associated tests from memory
+   * @param {string} locator   shortname/id, e.g. k/kayan
+   */
+  unregister: function(locator) {
+    let {shortname, id} = this.parseLocator(locator);
+    if(!id) throw new Error('Invalid locator '+locator);
+
+    keyman.removeKeyboards(id);
+    delete this.keyboards[id];
+  },
+
+  /**
+   * Run all tests for all loaded keyboards
+   */
+  runAllTests: function() {
+    for(var k in this.keyboards) {
+      this.runTests(k);
+    }
+  },
+
+  /**
+   * Run all tests for a given keyboard
+   * @param {string} id Keyboard identifier
+   */
+  runTests: function(keyboardId) {
+    return new Promise(function(resolve, reject) {
+      this.keyboards[keyboardId].results = {};
+      
+      console.log('-- Running '+this.keyboards[keyboardId].inputTests.length+' tests for '+keyboardId+'.');
+      var chunkSize = TEST_BATCH_SIZE;
+
+      // Batch tests into groups of chunkSize. This means we don't
+      // run into issues of the browser stalling on very long runs.
+
+      var tests = chunk(this.keyboards[keyboardId].inputTests, chunkSize), n = 0;
+      this.initProgress(tests.length);
+      var f = function() {
+        var testChunk = tests.shift();
+        testChunk.forEach((v) => this.runTest(keyboardId, v.id));
+        this.updateProgress(this.max - tests.length);
+        if(tests.length) {
+          window.setTimeout(f.bind(this), 1);
+        } else {
+          resolve(this.keyboards[keyboardId].results);
+        }
+      };
+
+      if(tests.length) window.setTimeout(f.bind(this), 1);
+    }.bind(this));
+  },
+
+  /**
+   * Run a single test for a keyboard
+   * @param {string}        id     Keyboard identifier
+   * @param {string|number} index  Test identifier
+   */
+  runTest: function(keyboardId, testId) {
+    var test = this.keyboards[keyboardId].inputTests.find((v) => v.id == testId);
+    if(test !== undefined) {
+      //console.log('Running test '+id+':'+index);
+      receiver.value = test.context || '';
+      //TODO: reset deadkey state
+      keyman.interface.resetContext();
+      var e = new com.keyman.KeyEvent();
+      e.Ltarg = receiver;
+      e.Lcode = test.key;
```

Comment:
So, unless I'm mistaken, only physical keystrokes are being tested at this point?

Reply:
Correct.